corehq/apps/sso/backends.py

```diff
@@ -0,0 +1,57 @@
+from django.contrib.auth.backends import ModelBackend
+from django.contrib.auth.models import User
+
+from corehq.apps.sso.models import IdentityProvider, AuthenticatedEmailDomain
+
+
+class SsoBackend(ModelBackend):
+    """
+    Authenticates against an IdentityProvider and SAML2 session data.
+    """
+
+    def authenticate(self, request, username=None, idp_slug=None, **kwargs):
+        #  Note: when implementing ODIC or other protocols in the future,
+        #   a different request.session check can be made
+        if not (username and idp_slug and request.session.get('samlSessionIndex')):
```

Comment:
Is the presence of `samlSessionIndex` in the session indicative of a successful SAML2 handshake?

Reply:
yes. And this is something I'm curious to open discussion about. `authenticate` using these specific `kwargs` would only be called with in the saml `ACS` view after the handshake is successful, and nowhere else. But should the `samlSessionIndex` be stored in postgres temporarily so that it can be cross-referenced here? My understanding is that unless someone has access to the `SECRET_KEY` the contents in `session` can't be spoofed. On top of it since the only trigger with all the required `kwargs` comes from within the `acs` that by itself adds security. no? am I incorrect in this understanding?